bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/test/java/com/google/cloud/bigtable/hbase/TestRetryBehavior.java

```diff
@@ -0,0 +1,297 @@
+          .isAtLeast(ATTEMPT_TIMEOUT.minus(Duration.ofMillis(10)));
+
+      // but should still be limited by the operation timeout with some jitter
+      assertThat(elapsed).isAtMost(OPERATION_TIMEOUT.plus(ATTEMPT_TIMEOUT.dividedBy(2)));
+    }
+  }
+
+  @Test
+  public void testRpcWillRespectAttemptTimeout() throws IOException {
+    rpcRunnable.set(
+        (call) -> {
+          /* hang indefinitely */
+        });
+
+    try (Connection connection = ConnectionFactory.createConnection(createBaseConfig());
+        Table table = connection.getTable(TableName.valueOf("fake-table"))) {
+
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      // TODO: this should be an IOException, but scan doesnt properly wrap it
+      Exception e = Assert.assertThrows(Exception.class, () -> testOp.runnable.run(table));
+
+      Duration elapsed = stopwatch.elapsed();
+
+      Status status = extractStatus(e);
+      assertThat(status.getCode()).isEqualTo(Status.Code.DEADLINE_EXCEEDED);
+
+      // Server will hang for the attempt timeout and the client should retry the attempt
+      // 5 * 450ms attempt timeouts > 2s operation timeout
+      assertThat(callCount.get()).isAtMost(5);
+      // 3 * 450ms attempt timeouts + jitter < 2s operation timeout
+      assertThat(callCount.get()).isAtLeast(3);
+      assertThat(Collections.max(remainingDurations)).isAtMost(ATTEMPT_TIMEOUT);
+
+      // Ignoring the first and last attempt, all attempts should be just under attempt timeout
+      assertThat(Collections.min(remainingDurations.subList(1, remainingDurations.size() - 1)))
+          .isAtLeast(ATTEMPT_TIMEOUT.minus(Duration.ofMillis(10)));
+
+      // Altogether should still be limited by the operation timeout with some jitter
+      assertThat(elapsed).isAtMost(OPERATION_TIMEOUT.plus(ATTEMPT_TIMEOUT.dividedBy(2)));
+    }
+  }
+
+  @Test
+  public void testRpcWillNotRetryOnHangWithoutAttemptTimeout() throws IOException {
+    rpcRunnable.set(
+        (call) -> {
+          /* hang indefinitely */
+        });
+
+    Configuration configuration = createBaseConfig();
+    setAttemptTimeout(configuration, OPERATION_TIMEOUT);
+
+    try (Connection connection = ConnectionFactory.createConnection(configuration);
+        Table table = connection.getTable(TableName.valueOf("fake-table"))) {
+
+      Stopwatch stopwatch = Stopwatch.createStarted();
+      // TODO: this should be an IOException, but scans dont wrap it properly
+      Exception e = Assert.assertThrows(Exception.class, () -> testOp.runnable.run(table));
+
+      Duration elapsed = stopwatch.elapsed();
+
+      Status status = extractStatus(e);
+      assertThat(status.getCode()).isEqualTo(Status.Code.DEADLINE_EXCEEDED);
+
+      // Server will hang for the attempt timeout and the client should retry the attempt
+      assertThat(callCount.get()).isEqualTo(1); // 5 * 450ms attempt timeouts > 2s operation timeout
```

Comment:
These 2 comments should be deleted:

// 5 * 450ms attempt timeouts > 2s operation timeout
// but should still be limited by the operation timeout with some jitter

Reply:
done